cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/AbstractInMemoryProgramRunner.java

```diff
@@ -130,17 +133,36 @@ private ProgramOptions createComponentOptions(String name, int instanceId, int i
     private final ProgramOptions options;
     private final Lock lock = new ReentrantLock();
     private final ProgramRunnerFactory.Type type;
+    private final AtomicLong liveComponents;
 
     public InMemoryProgramController(Table<String, Integer, ProgramController> components,
-                              RunId runId, Program program, ProgramSpecification spec,
-                              ProgramOptions options, ProgramRunnerFactory.Type type) {
+                                     RunId runId, Program program, ProgramSpecification spec,
+                                     ProgramOptions options, ProgramRunnerFactory.Type type) {
       super(program.getName(), runId);
       this.program = program;
       this.components = components;
       this.spec = spec;
       this.options = options;
       this.type = type;
+      this.liveComponents = new AtomicLong(components.size());
       started();
+      monitorComponents();
+    }
+
+    // Add listener to monitor completion/killed status of individual components, so that the program can be marked
+    // as completed once all the components have completed/killed.
+    private void monitorComponents() {
+      for (ProgramController controller : components.values()) {
+        controller.addListener(new AbstractListener() {
+          @Override
+          public void completed() {
+            if (liveComponents.decrementAndGet()  == 0) {
+              LOG.info("Program stopped: {}", spec.getName());
```

Comment:
@chtyim Added the LOG message in AbstractProgramController since we have the program started message there in started() method. Please take a look and let me know if that is not the right place.